### PR TITLE
#166240818 send the token to the user

### PR DESCRIPTION
**Pull request by alainburindi**
#### What does this PR do?
Send a JWT token to the user after successful registration or login
#### Description of Task to be completed?
* create user's model
* write unit tests
* allow a user to register
* allow a user to login
* save the token into the database
* send a JWT token
#### How should this be manually tested?
After cloning the repo, run 
* ```npm install``` 
* ```npm run dev``` 
* send the required data to ```localhost:3000/api/auth/signup``` to signup or
 * ```localhost:3000/api/auth/login``` to login
#### What are the relevant pivotal tracker stories?
[#166240818](https://www.pivotaltracker.com/story/show/166240818)

### CODE REVIEW

**Comment by murediane** [comment]
The test are failling, make the test pass please

**Review by murediane — COMMENTED** [review]
Code looks good to me however can you squash commit message please

**Comment by truestbyheart** [comment]
Try to put the manual instructions into a list of steps instead of a paragraph

**Review by truestbyheart — COMMENTED** [review]
Please use classes

**Comment by alainburindi** [comment]
> Code looks good to me however can you squash commit message please

done

**Comment by alainburindi** [comment]
> Try to put the manual instructions into a list of steps instead of a paragraph

check now

**Comment by MemunaHaruna** [comment]
@alainburindi please, ensure your branch name and commit messages match the conventions outlined here: 
https://github.com/andela/engineering-playbook/tree/c74e0b42578fcf45d76b91e43b85d5159821b7ea/5.%20Developing/Conventions#branch-naming  

For instance:
- your branch name shouldn't have the #
- squash your commit messages into one and ensure it follows the format specified in the engineering playbook.

File: src/api/controllers/authController.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+   * @returns {Boolean} true
+   */
+  static login(req, res) {
+    const { email, password } = req.body;
+    User.findByEmail(email).then((user) => {
+      if (user) {
+        const isPasswordValid = bcrypt.compareSync(password, user.dataValues.password);
+        if (isPasswordValid) {
+          const token = generateToken(user);
+          return sendResult(res, 200, 'user created successfully', user, token);
```
Comment (MemunaHaruna) [inline src/api/controllers/authController.js]: @alainburindi I think this should be a different message. Perhaps, 'user logged in successfully'

File: src/api/models/user.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+const user = (sequelize, DataTypes) => {
+  const User = sequelize.define('Users', {
+    names: {
```
Comment (MemunaHaruna) [inline src/api/models/user.js]: does this represent both the first name and last name? Is there a reason why they aren't separate fields in the DB?

File: src/configs/environnements.js [hunk 3]
```diff
@@ -4,27 +4,32 @@ dotenv.config();
 
 const port = process.env.PORT || 3000;
```
Comment (MemunaHaruna) [inline src/configs/environnements.js:5]: This filename should be 'environments.js'

File: src/helpers/results/send.auth.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+const sendResult = (res, status, message, user, token) => res.status(status).json({
+  message,
+  user: {
+    email: user.email,
+    token,
+    username: user.name
```
Comment (MemunaHaruna) [inline src/helpers/results/send.auth.js]: @alainburindi do we want to send this as 'username'? I'm thinking the name should reflect what the field represents. Except if we've made a design decision to refer to the user's name as username all through the app.

File: src/helpers/tokens/generate.token.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import environnements from '../../configs/environnements';
+
+const { Token } = models;
+
+const createToken = (user) => {
+  const token = jwt.sign(
+    { user },
+    environnements.currentEnv.secret,
+    {
+      expiresIn: '6h'
```
Comment (MemunaHaruna) [inline src/helpers/tokens/generate.token.js]: @alainburindi any reason why you've chosen 6h? Why not 24h? Just want to know your thoughts on this.

File: test/auth.test.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
+import 'babel-polyfill';
+import chai from 'chai';
+import chaiHttp from 'chai-http';
+import models from '../src/api/models/index';
+
+import server from '../src/index';
+
+// const should = chai.should();
```
Comment (MemunaHaruna) [inline test/auth.test.js]: please, remove unused code.

File: test/auth.test.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+      .post('/api/auth/signup')
+      .send({
+        names: 'Burindi Alain',
+        email: 'alain@gmail.com',
+        password: 'password',
+        confirmPassword: 'password'
+      })
+      .end((err, res) => {
+        res.should.have.status(201);
+        res.should.be.an('Object');
```
Comment (MemunaHaruna) [inline test/auth.test.js]: Nitpick: do we really care if it's an object? I think we should focus more on checking the content of the object returned.

File: test/auth.test.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+      .send({
+        email: 'alain@gmail.com',
+        password: 'password',
+      })
+      .end((err, res) => {
+        res.should.have.status(200);
+        res.body.should.be.an('Object');
+        res.body.should.have.property('user');
+        res.body.user.should.be.an('Object');
+        res.body.user.should.have.property('token');
```
Comment (MemunaHaruna) [inline test/auth.test.js:55]: Please, confirm that the right token, email etc are returned. This feedback applies to all the other tests.

File: test/auth.test.js [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+  it('should not login an unexisting user', (done) => {
+    chai.request(server)
+      .post('/api/auth/login')
+      .send({
+        email: 'alain666326@gmail.com',
+        password: 'password',
+      })
+      .end((err, res) => {
+        res.should.have.status(401);
```
Comment (MemunaHaruna) [inline test/auth.test.js]: Do we want to return a 401 here or a 404? Let me know what you think.

File: src/api/controllers/authController.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  static async signup(req, res) {
+    const { names, email, password } = req.body;
+    const salt = genSaltSync(parseFloat(process.env.BCRYPT_HASH_ROUNDS) || 10);
+    const hashedPassword = hashSync(password, salt);
+    const user = await User.create({
+      names,
+      email,
+      password: hashedPassword,
+    });
+    const token = generateToken(user);
```
Comment (MemunaHaruna) [inline src/api/controllers/authController.js]: @alainburindi if there's a server error while generating the token, are we catching that anywhere?

**Review by MemunaHaruna — CHANGES_REQUESTED** [review]

File: src/api/controllers/authController.js (hunk 1 above)
Comment (alainburindi) [reply src/api/controllers/authController.js]: done

File: src/configs/environnements.js (hunk 3 above)
Comment (premlogic) [reply src/configs/environnements.js:5]: I'm the one who wrote that name. Really sorry. Please @alainburindi change it

File: src/api/models/user.js (hunk 2 above)
Comment (alainburindi) [reply src/api/models/user.js]: it was a mistake according to the documentation I have to store the username, which is different from user's names. I'm going to change it

File: package.json [hunk 11]
```diff
@@ -5,7 +5,7 @@
   "main": "index.js",
```
Comment (premlogic) [inline package.json:5]: ```chai-http``` is in dependencies but also in dev dependencies. Can you remove it from dependencies @alainburindi

File: src/configs/environnements.js (hunk 3 above)
Comment (alainburindi) [reply src/configs/environnements.js:5]: it's done

File: src/api/controllers/authController.js [hunk 12]
```diff
@@ -0,0 +1,65 @@
+import bcrypt, { hashSync, genSaltSync } from 'bcrypt';
+import models from '../models/index';
+import generateToken from '../../helpers/tokens/generate.token';
+import sendResult from '../../helpers/results/send.auth';
+
+const { User } = models;
+
+/**
```
Comment (premlogic) [inline src/api/controllers/authController.js:8]: Can you please add a description of the functions in your ```JsDocs```?

**Review by nkpremices — CHANGES_REQUESTED** [review]

File: src/helpers/results/send.auth.js (hunk 4 above)
Comment (alainburindi) [reply src/helpers/results/send.auth.js]: I've made the changes to reflect the reality

File: src/helpers/tokens/generate.token.js (hunk 5 above)
Comment (alainburindi) [reply src/helpers/tokens/generate.token.js]: I thought that 24h is long

File: test/auth.test.js (hunk 6 above)
Comment (alainburindi) [reply test/auth.test.js]: done

File: src/helpers/tokens/generate.token.js (hunk 5 above)
Comment (MemunaHaruna) [reply src/helpers/tokens/generate.token.js]: Okay, please confirm this from the PO. Having users sign in every 6 hours seems inconvenient to me,  but maybe Aaron has a preference we can adopt.

File: test/auth.test.js (hunk 9 above)
Comment (alainburindi) [reply test/auth.test.js]: after discussing with my colleagues! it's a 4040

File: src/api/controllers/authController.js (hunk 10 above)
Comment (alainburindi) [reply src/api/controllers/authController.js]: done

File: package.json (hunk 11 above)
Comment (alainburindi) [reply package.json:5]: done

File: src/api/controllers/authController.js (hunk 12 above)
Comment (alainburindi) [reply src/api/controllers/authController.js:8]: done

File: src/api/controllers/authController.js [hunk 13]
```diff
@@ -0,0 +1,66 @@
+import bcrypt, { hashSync, genSaltSync } from 'bcrypt';
+import models from '../models/index';
+import generateToken from '../../helpers/tokens/generate.token';
+import sendResult from '../../helpers/results/send.auth';
+
+const { User } = models;
```
Comment (premlogic) [inline src/api/controllers/authController.js:6]: I think the users model should be in plural and shouldn't start with a capital letter. I would like to know the opinion of @MemunaHaruna

Comment (MemunaHaruna) [reply src/api/controllers/authController.js:6]: `User` is fine, in my opinion.

File: src/helpers/tokens/generate.token.js [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+      }
+    );
+
+    Token.create({
+      data: token,
+    });
+
+    return token;
+  } catch (error) {
+    return 'error while processing a token';
```
Comment (bdushimi) [inline src/helpers/tokens/generate.token.js:22]: @alainburindi, better to return the error message along with the error object in order to understand what went wrong.

File: test/auth.test.js (hunk 8 above)
Comment (MemunaHaruna) [reply test/auth.test.js:55]: @alainburindi please verify the other content of the response, not only the token.

e.g  
```
    user = res.body.user
    user.username.should.equal('Burindi Alain');
    user.email.should.equal('alain@gmail.com');
```

etc

It's important to thoroughly check that the right response is returned. 
N.b this applies to other tests, as well.

**Review by MemunaHaruna — COMMENTED** [review]
@alainburindi Please, let me know when you're done addressing the feedback so we can merge ASAP.

**Comment by alainburindi** [comment]
Screen Shot 2019-06-12 at 11 46 41


successful registration

**Comment by alainburindi** [comment]
Screen Shot 2019-06-12 at 11 47 32

successful login

**Comment by alainburindi** [comment]
Screen Shot 2019-06-12 at 11 47 51


user doesn't exist

**Comment by alainburindi** [comment]
Screen Shot 2019-06-12 at 11 48 11


password incorrect

File: src/api/controllers/authController.js (hunk 13 above)
Comment (premlogic) [reply src/api/controllers/authController.js:6]: Oh. Great. I will also follow that.

**Comment by murediane** [comment]
the build is failing, Please make them pass

**Comment by murediane** [comment]
Looks good to me

**Pull request merged**